### DISCUSSION TITLE
GUI Development

### DISCUSSION THREAD

**Discussion by roaldarbol** [BrianPugh/belay#112] [General]
Hi again Brian,
Still just trotting along slowly with my research here... this is a bit of a mind-ramble, but I hope you can follow and bear with me. :-)

I'm looking to incorporate Belay into Bux (my experimental GUI), and am trying to think of the best way to do it graphically/conceptually. I'm assuming that the user already has a folder with their Belay scripts, which has been tested to work.

My current idea is that the user can select a `Port` (board) from a dropdown that is generated by `belay.list_devices()` and an accompanying function. They choose the function by first selecting a Belay directory (which contains the needed scripts), then selects the script that they want, and lastly a function (maybe a bit convoluted and could possibly be done in fewer steps). This script could look something like this:
```python
from experiments.devices import Pico
import belay
import time

def run_of_sensor(port):
    pico = Pico.Pico(port)
    pico.import_libs()
    pico.setup_PAA5100(double=True)

    try:
        while True:
            pico.record_flow_double()
    except:
        pass
    pico.close()
```
...where the user would then choose `run_of_sensor()`.
The script that is run should also not really have other external CPython dependencies than `belay` (apart from possibly `numpy` and `pandas`).

I am intending for each device to run as a separate process (maybe a thread would do), and be interruptable by an event. From here on I'm also considering how the data should be saved. Should the function itself make use of a generator and save on the fly (maybe even take a destination as an input (`run_of_sensor(port, destination, event)`), or whether the GUI should carry out that part. 

Does that sound like it makes sense to you? Are there any blind angles I've missed here?

**Comment by BrianPugh** [comment]
Assuming that `Pico` inherits from `belay.Device`, I'd do the following:

```
def run_of_sensor(port):
    # use context manager so we don't have to worry about closing, and thusly catching exceptions.
    with Pico.Pico(port) as pico:  # you should fix your __init__.py so that this can just be ``Pico``
        pico.import_libs()  # this method can be decorated with the ``@setup(autoinit=True)`` so we don't need to call it here.
        pico.setup_PAA5100(double=True) 

        while True:
            pico.record_flow_double()
```

Belay will always call `close` (and, thusly, your `teardown` methods) on exit, so you won't need to worry about that too much (note: will obviously not happen if the device is physically disconnected).

**Reply by roaldarbol** [reply]
Thanks for the tips! Makes a lot of sense, I'll make some changes. And I've now got some code running that does the job.

By the way, I'm planning to make the Belay side of the GUI into a stand-alone app as well - don't know who it would be valuable to, but maybe someone. :-P
